# **Business Understanding**

*About data:*
This is a transactional data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail

*Note:*

InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If **this code starts with letter 'c', it indicates a cancellation.** 
StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each **distinct product**.
Description: Product (item) name. Nominal.
Quantity: The quantities of each product (item) per transaction. Numeric.	
InvoiceDate: Invoice Date and time. Numeric, the day and time when each transaction was generated.
UnitPrice: Unit price. Numeric, Product price per unit in sterling.
CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
Country: Country name. Nominal, the name of the country where each customer resides

*Source:*
Chen, D. (2015). Online Retail [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5BW33.
[source](https://archive.ics.uci.edu/dataset/352/online+retail)

#### **Load Libraries**

In [43]:
import pandas as pd 

#### **Load Data**

In [44]:
df = pd.read_excel("data/online+retail/Online Retail.xlsx")

# **Data Understanding**

#### **Overview of the data**

In [45]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [46]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France


In [47]:
rows, columns = df.shape
output = f"{rows} records and {columns} features"
print(output)

541909 records and 8 features


In [48]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

In [49]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


Insights:
- 541909 records and 8 features
- the data is composed of both numerical and categorial information together with datetime
- the data contains missing values

In [50]:
clean_df = df.drop("CustomerID", axis=1)

In [51]:
clean_df = df.drop("CustomerID", axis=1)

In [52]:
clean_df.describe() # numerical

,Quantity,InvoiceDate,UnitPrice
count,541909.000000,541909,541909.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114
min,-80995.000000,2010-12-01 08:26:00,-11062.060000
25%,1.000000,2011-03-28 11:34:00,1.250000
50%,3.000000,2011-07-19 17:17:00,2.080000
75%,10.000000,2011-10-19 11:27:00,4.130000
max,80995.000000,2011-12-09 12:50:00,38970.000000
std,218.081158,NaN,96.759853


In [53]:
difference = clean_df['InvoiceDate'].max() - clean_df['InvoiceDate'].min() 
print(difference)

373 days 04:24:00


In [ ]:
df[df["CustomerID"] == "John Doe"] # here we filter Rows whosecustomer ID is John Doe

In [54]:
df_r = clean_df[clean_df['Country'] == "Canada"]
df_r.head(2)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
119191,546533,20886,BOX OF 9 PEBBLE CANDLES,12,2011-03-14 13:53:00,1.95,Canada
119192,546533,79030D,"TUMBLER, BAROQUE",6,2011-03-14 13:53:00,1.65,Canada


In [ ]:
"""
- get the invoice number
- check if first letter is 'c' or it the invoce starts with c
- if it start we 'c' we can alert the team
"""

#### **Simulating canceled invoice since the data does not contain cancelled invoices**

In [69]:
clean_df.at[0, 'InvoiceNo'] = 'cc536365' # we have changed a cell value, whereby the cell is an intesection betweeb a row and column 

In [70]:
clean_df.head(2)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
0,cc536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,United Kingdom


In [62]:
clean_df["InvoiceNo"] = clean_df["InvoiceNo"].astype('str')

In [63]:
clean_df.dtypes

InvoiceNo                 str
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
Country                   str
dtype: object

In [64]:
clean_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,United Kingdom


In [71]:
cancelled_invoices = clean_df[clean_df["InvoiceNo"].str.startswith("c")]
cancelled_invoices

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
0,cc536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,United Kingdom


**Insights:**

*Quantity:*
- Majority of quantities selected were on avrage 9 items
- the lowst quanityt of items picked were 1
- the highest number was 80995**
- 25% of the order quanitties  had 1 order 
- 50% of the order quanitties were 3 and below
- 75% of the order quantities were 10 and below
- there existed a spread/ deviation from the mean by 218 items

*InvoiceDate:*
- 50% of the invoices were captured during 2011-07-19 17:17:00
- the first purchase/ invoice was done in 2010-12-01 08:26:00
- the last purchase/ invoice was done in 2011-12-09 12:50:00

In [ ]:
clean_df.describe(include='object')

/tmp/ipykernel_7030/1550864367.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  clean_df.describe(include='object')


,InvoiceNo,StockCode,Description,Country
count,541909,541909,540455,541909
unique,25900,4070,4223,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1114,2313,2369,495478


# **Data Preparation**

# **Modelling**

# **Evaluation**

# **Deployment**